File: src/scanner.js
const { ethers } = require('ethers');
const chalk = require('chalk');
const Table = require('cli-table3');

// Configuration
const CONFIG = {
    ETH_PRICE: 4300,
    GAS_PRICE_GWEI: 0.5,
    GAS_LIMIT: 600000,
    MIN_LIQUIDITY: 1000, // Minimum liquidity in USD
    MIN_SPREAD: 0.01, // 0.01% minimum spread to show
};

// RPC Endpoints (all free, no API key needed)
const RPC_ENDPOINTS = [
    'https://eth.llamarpc.com',
    'https://ethereum.publicnode.com',
    'https://1rpc.io/eth',
    'https://eth.drpc.org',
    'https://rpc.payload.de',
    'https://eth-mainnet.public.blastapi.io',
    'https://rpc.ankr.com/eth',
    'https://cloudflare-eth.com',
];

// DEX Factory Addresses
const FACTORIES = {
    'UniswapV2': '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f',
    'SushiSwap': '0xC0AEe478e3658e2610c5F7A4A2E1777cE9e4f2Ac',
    'ShibaSwap': '0x115934131916C8b277DD010Ee02de363c09d037c',
    'FraxSwap': '0x43eC799eAdd63848443E2347C49f5f52e8Fe0F6f',
    'PancakeV2': '0x1097053Fd2ea711dad45caCcc45EfF7548fCB362',
};

// Flash Loan Providers
const FLASH_PROVIDERS = [
    { name: 'Balancer', fee: 0.0000, color: 'green' },    // 0% fee!
    { name: 'dYdX', fee: 0.0000, color: 'green' },        // 2 wei (basically 0%)
    { name: 'UniswapV3', fee: 0.0001, color: 'yellow' },  // 0.01%
    { name: 'Aave V3', fee: 0.0005, color: 'yellow' },    // 0.05%
];

// ABIs
const FACTORY_ABI = [
    'function allPairsLength() view returns (uint256)',
    'function allPairs(uint256) view returns (address)',
    'function getPair(address, address) view returns (address)',
];

const PAIR_ABI = [
    'function getReserves() view returns (uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast)',
    'function token0() view returns (address)',
    'function token1() view returns (address)',
];

const ERC20_ABI = [
    'function symbol() view returns (string)',
    'function decimals() view returns (uint8)',
    'function name() view returns (string)',
];

class ArbitrageScanner {
    constructor() {
        this.provider = null;
        this.tokenCache = new Map();
        this.pairCache = new Map();
        this.opportunities = [];
        this.totalScanned = 0;
        this.profitableCount = 0;
        this.totalPotentialProfit = 0;
    }

    async initialize() {
        console.log(chalk.cyan('\n════════════════════════════════════════════════════════════════════'));
        console.log(chalk.cyan.bold('   💎 ETHEREUM ARBITRAGE SCANNER - EVERY PAIR, EVERY OPPORTUNITY 💎'));
        console.log(chalk.cyan('════════════════════════════════════════════════════════════════════\n'));

        // Connect to RPC
        for (const endpoint of RPC_ENDPOINTS) {
            try {
                this.provider = new ethers.JsonRpcProvider(endpoint);
                await this.provider.getBlockNumber();
                console.log(chalk.green(`✅ Connected to ${endpoint}`));
                break;
            } catch (error) {
                continue;
            }
        }

        if (!this.provider) {
            throw new Error('Could not connect to any RPC endpoint');
        }

        // Get current network state
        const [blockNumber, gasPrice] = await Promise.all([
            this.provider.getBlockNumber(),
            this.provider.getFeeData()
        ]);

        const gasPriceGwei = Number(gasPrice.gasPrice) / 1e9;
        const gasUsd = (CONFIG.GAS_LIMIT * gasPriceGwei / 1e9) * CONFIG.ETH_PRICE;

        console.log(chalk.yellow('\n📊 Market Conditions:'));
        console.log(`  • Block: #${blockNumber}`);
        console.log(`  • Gas: ${gasPriceGwei.toFixed(2)} gwei ($${gasUsd.toFixed(2)}/tx)`);
        console.log(`  • ETH Price: $${CONFIG.ETH_PRICE}`);
        
        console.log(chalk.magenta('\n🏦 Flash Loan Providers:'));
        FLASH_PROVIDERS.forEach(p => {
            const fee = p.fee * 100;
            console.log(`  • ${p.name}: ${fee}% fee`);
        });
    }

    async scanAllPairs() {
        console.log(chalk.cyan('\n🔍 Scanning ALL pairs across ALL DEXs...\n'));

        const allPairs = new Map(); // token0-token1 -> [pools]

        // Scan each DEX
        for (const [dexName, factoryAddress] of Object.entries(FACTORIES)) {
            console.log(chalk.yellow(`\n📍 Scanning ${dexName}...`));
            
            const factory = new ethers.Contract(factoryAddress, FACTORY_ABI, this.provider);
            
            try {
                const totalPairs = await factory.allPairsLength();
                console.log(`   Found ${totalPairs} total pairs`);

                // Sample pairs (get recent + random for speed)
                const indices = this.getSampleIndices(Number(totalPairs), 100);
                
                for (const index of indices) {
                    try {
                        const pairAddress = await factory.allPairs(index);
                        const poolData = await this.getPoolData(pairAddress, dexName);
                        
                        if (poolData && this.hasMinimumLiquidity(poolData)) {
                            const key = this.getPairKey(poolData.token0, poolData.token1);
                            
                            if (!allPairs.has(key)) {
                                allPairs.set(key, []);
                            }
                            allPairs.get(key).push(poolData);
                            this.totalScanned++;
                        }
                    } catch (error) {
                        // Skip failed pairs
                    }
                }
            } catch (error) {
                console.log(chalk.red(`   Error scanning ${dexName}: ${error.message}`));
            }
        }

        console.log(chalk.green(`\n✅ Scanned ${this.totalScanned} pools total`));
        console.log(chalk.green(`✅ Found ${allPairs.size} unique token pairs\n`));

        // Find arbitrage opportunities
        await this.findArbitrageOpportunities(allPairs);
    }

    getSampleIndices(total, sampleSize) {
        const indices = [];
        
        // Get last 50 pairs (most recent/active)
        const start = Math.max(0, total - 50);
        for (let i = start; i < total && indices.length < sampleSize / 2; i++) {
            indices.push(i);
        }
        
        // Get random samples from the rest
        const remaining = sampleSize - indices.length;
        for (let i = 0; i < remaining && i < total; i++) {
            const randomIndex = Math.floor(Math.random() * Math.min(total, 10000));
            if (!indices.includes(randomIndex)) {
                indices.push(randomIndex);
            }
        }
        
        return indices;
    }

    async getPoolData(pairAddress, dexName) {
        try {
            const pair = new ethers.Contract(pairAddress, PAIR_ABI, this.provider);
            
            const [reserves, token0, token1] = await Promise.all([
                pair.getReserves(),
                pair.token0(),
                pair.token1()
            ]);

            // Get token info
            const [token0Info, token1Info] = await Promise.all([
                this.getTokenInfo(token0),
                this.getTokenInfo(token1)
            ]);

            return {
                dex: dexName,
                address: pairAddress,
                token0,
                token1,
                symbol0: token0Info.symbol,
                symbol1: token1Info.symbol,
                decimals0: token0Info.decimals,
                decimals1: token1Info.decimals,
                reserve0: reserves.reserve0,
                reserve1: reserves.reserve1,
            };
        } catch (error) {
            return null;
        }
    }

    async getTokenInfo(tokenAddress) {
        if (this.tokenCache.has(tokenAddress)) {
            return this.tokenCache.get(tokenAddress);
        }

        try {
            const token = new ethers.Contract(tokenAddress, ERC20_ABI, this.provider);
            const [symbol, decimals] = await Promise.all([
                token.symbol().catch(() => 'UNKNOWN'),
                token.decimals().catch(() => 18)
            ]);

            const info = { symbol, decimals };
            this.tokenCache.set(tokenAddress, info);
            return info;
        } catch (error) {
            const info = { symbol: 'UNKNOWN', decimals: 18 };
            this.tokenCache.set(tokenAddress, info);
            return info;
        }
    }

    hasMinimumLiquidity(pool) {
        const reserve0 = Number(pool.reserve0) / Math.pow(10, pool.decimals0);
        const reserve1 = Number(pool.reserve1) / Math.pow(10, pool.decimals1);
        
        // Rough USD estimate (assuming at least one token is worth ~$1)
        const estimatedValue = Math.max(reserve0, reserve1);
        return estimatedValue > CONFIG.MIN_LIQUIDITY;
    }

    getPairKey(token0, token1) {
        return token0.toLowerCase() < token1.toLowerCase() 
            ? `${token0}-${token1}` 
            : `${token1}-${token0}`;
    }

    async findArbitrageOpportunities(pairMap) {
        console.log(chalk.cyan('🎯 Analyzing arbitrage opportunities...\n'));

        for (const [pairKey, pools] of pairMap.entries()) {
            if (pools.length < 2) continue;

            // Check all DEX combinations
            for (let i = 0; i < pools.length; i++) {
                for (let j = i + 1; j < pools.length; j++) {
                    const opportunity = this.calculateArbitrage(pools[i], pools[j]);
                    if (opportunity) {
                        this.opportunities.push(opportunity);
                        this.displayOpportunity(opportunity);
                    }
                }
            }
        }

        this.displaySummary();
    }

    calculateArbitrage(pool1, pool2) {
        // Calculate prices
        const price1 = this.calculatePrice(pool1);
        const price2 = this.calculatePrice(pool2);

        if (!price1 || !price2) return null;

        // Calculate spread
        const spread = Math.abs(price1 - price2) / Math.min(price1, price2) * 100;

        if (spread < CONFIG.MIN_SPREAD) return null;

        const buyPool = price1 < price2 ? pool1 : pool2;
        const sellPool = price1 < price2 ? pool2 : pool1;
        const buyPrice = Math.min(price1, price2);
        const sellPrice = Math.max(price1, price2);

        // Calculate profitability for different loan sizes
        const calculations = [];
        const loanSizes = [0.1, 1, 10, 100, 1000];
        
        for (const loanEth of loanSizes) {
            const loanUsd = loanEth * CONFIG.ETH_PRICE;
            const grossProfit = loanUsd * spread / 100;
            
            // Find best flash loan provider
            let bestProvider = null;
            let bestNetProfit = -Infinity;
            
            for (const provider of FLASH_PROVIDERS) {
                const flashFee = loanUsd * provider.fee;
                const dexFees = loanUsd * 0.006; // 0.3% each way
                const gasUsd = (CONFIG.GAS_LIMIT * CONFIG.GAS_PRICE_GWEI / 1e9) * CONFIG.ETH_PRICE;
                const netProfit = grossProfit - flashFee - dexFees - gasUsd;
                
                if (netProfit > bestNetProfit) {
                    bestNetProfit = netProfit;
                    bestProvider = provider;
                }
            }
            
            calculations.push({
                loanEth,
                loanUsd,
                grossProfit,
                netProfit: bestNetProfit,
                provider: bestProvider,
                profitable: bestNetProfit > 0
            });
        }

        // Find best profitable opportunity
        const bestCalc = calculations
            .filter(c => c.profitable)
            .sort((a, b) => b.netProfit - a.netProfit)[0];

        if (!bestCalc) return null;

        this.profitableCount++;
        this.totalPotentialProfit += bestCalc.netProfit;

        return {
            pair: `${buyPool.symbol0}/${buyPool.symbol1}`,
            buyDex: buyPool.dex,
            sellDex: sellPool.dex,
            spread,
            buyPrice,
            sellPrice,
            calculations,
            bestCalc,
            timestamp: new Date().toLocaleTimeString()
        };
    }

    calculatePrice(pool) {
        const reserve0 = Number(pool.reserve0) / Math.pow(10, pool.decimals0);
        const reserve1 = Number(pool.reserve1) / Math.pow(10, pool.decimals1);
        
        if (reserve0 === 0) return null;
        return reserve1 / reserve0;
    }

    displayOpportunity(opp) {
        const profitColor = opp.bestCalc.netProfit > 100 ? 'green' : 
                           opp.bestCalc.netProfit > 10 ? 'yellow' : 'white';

        console.log(chalk.bgMagenta.bold(`\n 💰 ARBITRAGE OPPORTUNITY #${this.profitableCount} `));
        console.log(chalk.white('═'.repeat(70)));
        
        console.log(chalk.cyan(`📍 Pair: ${opp.pair}`));
        console.log(chalk.cyan(`🔄 Route: ${opp.buyDex} → ${opp.sellDex}`));
        console.log(chalk.yellow(`📊 Spread: ${opp.spread.toFixed(4)}%`));
        console.log(chalk.gray(`⏰ Time: ${opp.timestamp}`));
        
        // Show profitability table
        const table = new Table({
            head: ['Loan Size', 'Flash Provider', 'Gross Profit', 'Net Profit', 'Status'],
            style: { head: ['cyan'] }
        });

        opp.calculations.forEach(calc => {
            if (calc.profitable) {
                table.push([
                    `${calc.loanEth} ETH`,
                    calc.provider.name,
                    `$${calc.grossProfit.toFixed(2)}`,
                    chalk[profitColor](`$${calc.netProfit.toFixed(2)}`),
                    chalk.green('✅ PROFITABLE')
                ]);
            }
        });

        console.log(table.toString());
        
        console.log(chalk.bgGreen.black.bold(` 🎯 BEST: ${opp.bestCalc.loanEth} ETH loan via ${opp.bestCalc.provider.name} = $${opp.bestCalc.netProfit.toFixed(2)} profit `));
        console.log(chalk.white('═'.repeat(70)));
    }

    displaySummary() {
        console.log(chalk.cyan('\n════════════════════════════════════════════════════════════════════'));
        console.log(chalk.cyan.bold('                           📊 SUMMARY                              '));
        console.log(chalk.cyan('════════════════════════════════════════════════════════════════════'));
        
        console.log(chalk.yellow(`\n📈 Statistics:`));
        console.log(`  • Total Pools Scanned: ${this.totalScanned}`);
        console.log(`  • Profitable Opportunities: ${this.profitableCount}`);
        console.log(`  • Total Potential Profit: ${chalk.green.bold(`$${this.totalPotentialProfit.toFixed(2)}`)}`);
        
        if (this.profitableCount > 0) {
            console.log(chalk.green.bold(`\n✅ Found ${this.profitableCount} profitable arbitrage opportunities!`));
            console.log(chalk.yellow('💡 Execute these with flash loans for risk-free profit!'));
        }
    }

    async run() {
        await this.initialize();
        
        while (true) {
            this.opportunities = [];
            this.totalScanned = 0;
            this.profitableCount = 0;
            this.totalPotentialProfit = 0;
            
            await this.scanAllPairs();
            
            console.log(chalk.gray('\n⏳ Waiting 30 seconds before next scan...'));
            await new Promise(resolve => setTimeout(resolve, 30000));
        }
    }
}

// Main execution
async function main() {
    const scanner = new ArbitrageScanner();
    
    try {
        await scanner.run();
    } catch (error) {
        console.error(chalk.red('Error:', error.message));
        process.exit(1);
    }
}

// Run the scanner
main().catch(console.error);

File: src/scanner.rs
use anyhow::Result;
use chrono::Local;
use ethers::types::U256;
use log::{error, info, warn, debug};
use std::sync::Arc;
use tokio::time::{sleep, Duration};

use crate::config::Config;
use crate::dex::DexManager;
use crate::flashloan::FlashLoanManager;
use crate::gas::GasEstimator;
use crate::models::{ArbitrageOpportunity, DexPool, FlashLoanProvider};
use crate::providers::MultiProvider;

pub struct ArbitrageScanner {
    config: Arc<Config>,
    provider: Arc<MultiProvider>,
    dex_manager: DexManager,
    flash_loan_manager: FlashLoanManager,
    gas_estimator: GasEstimator,
}

impl ArbitrageScanner {
    pub async fn new(config: Arc<Config>, provider: Arc<MultiProvider>) -> Result<Self> {
        let dex_manager = DexManager::new(provider.clone()).await?;
        let flash_loan_manager = FlashLoanManager::new(
            provider.clone(),
            config.aave_v3_pool,
            config.balancer_vault,
        );
        let gas_estimator = GasEstimator::new(provider.clone()).await?;

        Ok(Self {
            config,
            provider,
            dex_manager,
            flash_loan_manager,
            gas_estimator,
        })
    }

    pub async fn run(&self) -> Result<()> {
        let mut iteration = 0u64;
        let mut total_opportunities_found = 0u64;
        let mut profitable_opportunities_found = 0u64;
        
        println!("\n🚀 ARBITRAGE SCANNER STARTED");
        println!("================================");
        println!("Settings:");
        println!("  Min Profit: ${}", self.config.min_profit_usd);
        println!("  Max Gas: {} gwei", self.config.max_gas_price_gwei);
        println!("  Scan Interval: {}ms", self.config.scan_interval_ms);
        println!("================================\n");

        loop {
            iteration += 1;
            
            match self.scan_cycle(iteration).await {
                Ok(opportunities) => {
                    if !opportunities.is_empty() {
                        total_opportunities_found += opportunities.len() as u64;
                        
                        // Count profitable ones
                        let profitable = opportunities.iter()
                            .filter(|o| o.net_profit_usd > 0.0)
                            .count() as u64;
                        profitable_opportunities_found += profitable;
                        
                        self.display_opportunities(&opportunities);
                        
                        // Show statistics
                        println!("\n📊 STATISTICS:");
                        println!("  Total opportunities found: {}", total_opportunities_found);
                        println!("  Profitable opportunities: {}", profitable_opportunities_found);
                        println!("  Current scan: #{}", iteration);
                    } else if iteration % 10 == 0 {
                        println!("⏳ Scan #{}: No opportunities (checked {} times, found {} total, {} profitable)", 
                            iteration, iteration, total_opportunities_found, profitable_opportunities_found);
                    }
                }
                Err(e) => {
                    error!("Scan cycle error: {}", e);
                }
            }

            sleep(Duration::from_millis(self.config.scan_interval_ms)).await;
        }
    }

    async fn scan_cycle(&self, iteration: u64) -> Result<Vec<ArbitrageOpportunity>> {
        // Update gas price and ETH price periodically
        if iteration % 5 == 0 {
            if let Err(e) = self.gas_estimator.update_eth_price().await {
                warn!("Failed to update ETH price: {}", e);
            }
            
            if let Ok(gas_price) = self.gas_estimator.get_current_gas_price().await {
                debug!("Current gas price: {:.2} gwei", gas_price.total_gwei);
                
                if gas_price.total_gwei > self.config.max_gas_price_gwei as f64 {
                    warn!("⚠️ Gas too high: {:.2} gwei > {} gwei (opportunities will be less profitable)",
                        gas_price.total_gwei, self.config.max_gas_price_gwei
                    );
                }
            }
        }

        // Get current block
        let block_number = self.provider.get_block_number().await?;
        
        if iteration == 1 || iteration % 20 == 0 {
            info!("Scanning block #{}", block_number);
        }

        // Fetch all DEX pools
        let pools = self.dex_manager.get_all_pools().await?;
        
        if pools.is_empty() {
            warn!("No pools found");
            return Ok(Vec::new());
        }

        // Find arbitrage opportunities
        let raw_opportunities = self.find_flash_loan_arbitrage(&pools);
        
        if !raw_opportunities.is_empty() {
            println!("🔍 Found {} potential arbitrage paths, calculating actual profits...", 
                raw_opportunities.len());
            
            // Show sample of what we found
            for (i, (buy, sell, _amount, diff)) in raw_opportunities.iter().take(5).enumerate() {
                println!("  #{}: {}/{} - Buy on {} → Sell on {} (spread: {:.2}%)",
                    i + 1,
                    buy.token_pair.symbol0,
                    buy.token_pair.symbol1,
                    buy.dex,
                    sell.dex,
                    diff.as_u64() as f64 / 100.0
                );
            }
        }
        
        // Limit to top 20 opportunities for speed
        let top_opportunities: Vec<_> = raw_opportunities.into_iter()
            .take(20)
            .collect();
        
        // Calculate ACTUAL profitability with all fees
        let mut all_opportunities = Vec::new();
        
        for (buy_pool, sell_pool, borrow_amount, _) in top_opportunities {
            // Test with Balancer first (0% fee)
            if let Ok(opportunity) = self
                .calculate_accurate_profit(
                    buy_pool.clone(),
                    sell_pool.clone(),
                    borrow_amount,
                    FlashLoanProvider::Balancer,
                    block_number,
                )
                .await
            {
                all_opportunities.push(opportunity);
            }
        }
        
        // Sort by net profit
        all_opportunities.sort_by(|a, b| b.net_profit_usd.partial_cmp(&a.net_profit_usd).unwrap());
        
        Ok(all_opportunities)
    }

    fn find_flash_loan_arbitrage(&self, pools: &[DexPool]) -> Vec<(DexPool, DexPool, U256, U256)> {
        let mut opportunities = Vec::new();

        // Group pools by token pair
        let mut pool_map: std::collections::HashMap<(Address, Address), Vec<&DexPool>> =
            std::collections::HashMap::new();

        use ethers::types::Address;
        for pool in pools {
            let key = if pool.token_pair.token0 < pool.token_pair.token1 {
                (pool.token_pair.token0, pool.token_pair.token1)
            } else {
                (pool.token_pair.token1, pool.token_pair.token0)
            };
            pool_map.entry(key).or_insert_with(Vec::new).push(pool);
        }

        debug!("Analyzing {} unique token pairs", pool_map.len());

        // Find arbitrage opportunities
        for (_pair, pools_for_pair) in pool_map.iter() {
            if pools_for_pair.len() < 2 {
                continue;
            }

            for i in 0..pools_for_pair.len() {
                for j in i + 1..pools_for_pair.len() {
                    let pool1 = pools_for_pair[i];
                    let pool2 = pools_for_pair[j];

                    // Prioritize cross-DEX opportunities
                    if pool1.dex == pool2.dex {
                        continue;
                    }

                    // Calculate price difference
                    let price1 = self.calculate_price(pool1);
                    let price2 = self.calculate_price(pool2);
                    
                    if price1.is_zero() || price2.is_zero() {
                        continue;
                    }

                    // Calculate percentage difference
                    let (lower_price, higher_price, buy_pool, sell_pool) = 
                        if price1 < price2 {
                            (price1, price2, pool1, pool2)
                        } else {
                            (price2, price1, pool2, pool1)
                        };

                    let price_diff = higher_price - lower_price;
                    let diff_percentage = (price_diff * U256::from(10000)) / lower_price;

                    // Need at least 0.65% to cover both DEX fees (0.3% each)
                    if diff_percentage > U256::from(65) {
                        // Calculate optimal borrow amount
                        let borrow_amount = self.calculate_optimal_borrow(buy_pool, sell_pool);
                        
                        if borrow_amount > U256::from(10u128.pow(16)) { // Min 0.01 tokens
                            opportunities.push((
                                (*buy_pool).clone(),
                                (*sell_pool).clone(),
                                borrow_amount,
                                diff_percentage
                            ));
                            
                            debug!("Found opportunity: {}/{} - {:.2}% spread between {} and {}", 
                                buy_pool.token_pair.symbol0,
                                buy_pool.token_pair.symbol1,
                                diff_percentage.as_u64() as f64 / 100.0,
                                buy_pool.dex,
                                sell_pool.dex
                            );
                        }
                    }
                }
            }
        }

        opportunities
    }

    fn calculate_optimal_borrow(&self, buy_pool: &DexPool, sell_pool: &DexPool) -> U256 {
        // Use the actual optimal arbitrage formula
        // For simplicity, using a conservative approach: 0.5% of smaller reserve
        let smaller_reserve = buy_pool.reserve0.min(sell_pool.reserve0);
        let optimal = smaller_reserve / U256::from(200);
        
        // Cap at 100 ETH worth to be realistic
        let max_borrow = U256::from(100u128 * 10u128.pow(18));
        optimal.min(max_borrow)
    }

    async fn calculate_accurate_profit(
        &self,
        buy_pool: DexPool,
        sell_pool: DexPool,
        borrow_amount: U256,
        flashloan_provider: FlashLoanProvider,
        block_number: u64,
    ) -> Result<ArbitrageOpportunity> {
        // Step 1: Calculate the arbitrage trade path
        // Borrow token0 -> Buy token1 on buy_pool -> Sell token1 on sell_pool -> Get token0 back
        
        // Calculate first swap (token0 -> token1 on buy_pool)
        let token1_received = self.dex_manager.calculate_output_amount(
            borrow_amount,
            buy_pool.reserve0,
            buy_pool.reserve1,
            buy_pool.fee,
        );
        
        // Calculate second swap (token1 -> token0 on sell_pool)
        let token0_received = self.dex_manager.calculate_output_amount(
            token1_received,
            sell_pool.reserve1,
            sell_pool.reserve0,
            sell_pool.fee,
        );
        
        // Step 2: Calculate flash loan costs
        let flash_loan_fee_bps = self
            .flash_loan_manager
            .get_flash_loan_fee(flashloan_provider)
            .await?;
        
        let flash_loan_fee = self
            .flash_loan_manager
            .calculate_flash_loan_cost(borrow_amount, flash_loan_fee_bps);
        
        // Total amount we need to repay (principal + fee)
        let repay_amount = borrow_amount + flash_loan_fee;
        
        // Step 3: Calculate gross profit
        let gross_profit_wei = if token0_received > repay_amount {
            token0_received - repay_amount
        } else {
            U256::zero()
        };
        
        // Step 4: Calculate gas costs
        let gas_estimate = self.gas_estimator.estimate_arbitrage_gas().await?;
        
        // Add extra gas for flash loan operations
        let flash_loan_extra_gas = U256::from(100000); // Simplified - was calling undefined method
        let total_gas = gas_estimate.gas_limit + flash_loan_extra_gas;
        let total_gas_cost_wei = total_gas * (gas_estimate.gas_price.base_fee + gas_estimate.gas_price.priority_fee);
        let total_gas_cost_usd = self.gas_estimator.wei_to_usd(total_gas_cost_wei).await;
        
        // Step 5: Calculate final NET profit - FIX THE AWAIT HERE
        let gross_profit_usd = self.gas_estimator.wei_to_usd(gross_profit_wei).await;
        let net_profit_usd = gross_profit_usd - total_gas_cost_usd;
        
        // Step 6: Calculate additional metrics
        let _price_impact = self.estimate_price_impact(&buy_pool, &sell_pool, borrow_amount);
        
        // Log the calculation breakdown
        if net_profit_usd > 0.0 || gross_profit_usd > 10.0 {
            debug!("📊 Profit Breakdown:");
            debug!("  Borrow: {} {}", 
                crate::utils::format_token_amount(borrow_amount, buy_pool.token_pair.decimals0),
                buy_pool.token_pair.symbol0
            );
            debug!("  After swap 1: {} {}", 
                crate::utils::format_token_amount(token1_received, buy_pool.token_pair.decimals1),
                buy_pool.token_pair.symbol1
            );
            debug!("  After swap 2: {} {}", 
                crate::utils::format_token_amount(token0_received, buy_pool.token_pair.decimals0),
                buy_pool.token_pair.symbol0
            );
            debug!("  Flash loan fee: {} {} ({} bps)", 
                crate::utils::format_token_amount(flash_loan_fee, buy_pool.token_pair.decimals0),
                buy_pool.token_pair.symbol0,
                flash_loan_fee_bps
            );
            debug!("  Gross profit: ${:.2}", gross_profit_usd);
            debug!("  Gas cost: ${:.2}", total_gas_cost_usd);
            debug!("  NET PROFIT: ${:.2}", net_profit_usd);
        }

        Ok(ArbitrageOpportunity {
            token_pair: buy_pool.token_pair.clone(),
            buy_pool,
            sell_pool,
            optimal_amount: borrow_amount,
            profit_wei: gross_profit_wei,
            profit_usd: gross_profit_usd,
            gas_cost_wei: total_gas_cost_wei,
            gas_cost_usd: total_gas_cost_usd,
            net_profit_usd,
            flashloan_provider,
            block_number,
        })
    }

    fn estimate_price_impact(&self, buy_pool: &DexPool, sell_pool: &DexPool, amount: U256) -> f64 {
        // Estimate how much our trade will move the price
        // Simplified calculation
        let impact_buy = amount.as_u128() as f64 / buy_pool.reserve0.as_u128() as f64;
        let impact_sell = amount.as_u128() as f64 / sell_pool.reserve0.as_u128() as f64;
        (impact_buy + impact_sell) * 100.0 // Return as percentage
    }

    fn calculate_price(&self, pool: &DexPool) -> U256 {
        if pool.reserve0.is_zero() || pool.reserve1.is_zero() {
            return U256::zero();
        }
        (pool.reserve1 * U256::from(10u128.pow(18))) / pool.reserve0
    }

    fn display_opportunities(&self, opportunities: &[ArbitrageOpportunity]) {
        println!("\n{}", "=".repeat(80));
        println!("💰 ARBITRAGE OPPORTUNITIES FOUND - {}", Local::now().format("%Y-%m-%d %H:%M:%S"));
        println!("{}", "=".repeat(80));

        let mut profitable = Vec::new();
        let mut unprofitable = Vec::new();

        for opp in opportunities {
            if opp.net_profit_usd > 0.0 {
                profitable.push(opp);
            } else {
                unprofitable.push(opp);
            }
        }

        // Show profitable opportunities first
        if !profitable.is_empty() {
            println!("\n✅ PROFITABLE OPPORTUNITIES ({} found):", profitable.len());
            println!("{}", "-".repeat(80));
            
            for (i, opp) in profitable.iter().enumerate() {
                println!("\n🎯 Opportunity #{} [PROFITABLE]", i + 1);
                self.display_opportunity_details(opp);
            }
        }

        // Show unprofitable ones for learning
        if !unprofitable.is_empty() {
            println!("\n❌ UNPROFITABLE OPPORTUNITIES ({} found):", unprofitable.len());
            println!("{}", "-".repeat(80));
            
            for (i, opp) in unprofitable.iter().take(3).enumerate() {
                println!("\n📉 Opportunity #{} [LOSS]", profitable.len() + i + 1);
                self.display_opportunity_details(opp);
            }
            
            if unprofitable.len() > 3 {
                println!("\n... and {} more unprofitable opportunities", unprofitable.len() - 3);
            }
        }

        // Summary
        println!("\n{}", "=".repeat(80));
        println!("📊 SUMMARY:");
        println!("  Total opportunities analyzed: {}", opportunities.len());
        println!("  Profitable: {} ({}%)", 
            profitable.len(), 
            (profitable.len() * 100) / opportunities.len().max(1)
        );
        println!("  Unprofitable: {}", unprofitable.len());
        
        if !profitable.is_empty() {
            let total_profit: f64 = profitable.iter().map(|o| o.net_profit_usd).sum();
            let best_profit = profitable.iter()
                .map(|o| o.net_profit_usd)
                .max_by(|a, b| a.partial_cmp(b).unwrap())
                .unwrap_or(0.0);
            
            println!("  Best opportunity: ${:.2}", best_profit);
            println!("  Total potential profit: ${:.2}", total_profit);
        }
        
        println!("{}", "=".repeat(80));
    }

    fn display_opportunity_details(&self, opp: &ArbitrageOpportunity) {
        println!("  📍 Token Pair: {}/{}", opp.token_pair.symbol0, opp.token_pair.symbol1);
        println!("  🔄 Route: {} → {} → {}", 
            opp.flashloan_provider,
            opp.buy_pool.dex,
            opp.sell_pool.dex
        );
        
        println!("\n  💵 FINANCIAL BREAKDOWN:");
        println!("  ├─ Flash Loan Amount: {} {}", 
            crate::utils::format_token_amount(opp.optimal_amount, opp.token_pair.decimals0),
            opp.token_pair.symbol0
        );
        println!("  ├─ Gross Profit: ${:.4} ({} wei)", 
            opp.profit_usd,
            opp.profit_wei
        );
        println!("  ├─ Gas Cost: ${:.4}", opp.gas_cost_usd);
        
        if opp.net_profit_usd > 0.0 {
            println!("  └─ NET PROFIT: ${:.4} ✅", opp.net_profit_usd);
        } else {
            println!("  └─ NET LOSS: ${:.4} ❌", opp.net_profit_usd);
        }
        
        // Show profitability metrics
        let roi = if opp.gas_cost_usd > 0.0 {
            (opp.net_profit_usd / opp.gas_cost_usd) * 100.0
        } else {
            0.0
        };
        
        println!("\n  📈 METRICS:");
        println!("  ├─ ROI: {:.1}%", roi);
        println!("  ├─ Gas/Profit Ratio: {:.1}%", 
            if opp.profit_usd > 0.0 {
                (opp.gas_cost_usd / opp.profit_usd) * 100.0
            } else {
                100.0
            }
        );
        println!("  └─ Block: #{}", opp.block_number);
        
        // Execution instructions if profitable
        if opp.net_profit_usd > 0.0 {
            println!("\n  ⚡ EXECUTION STEPS:");
            println!("  1. Flash borrow {} {} from {}", 
                crate::utils::format_token_amount(opp.optimal_amount, opp.token_pair.decimals0),
                opp.token_pair.symbol0,
                opp.flashloan_provider
            );
            println!("  2. Swap to {} on {} (at lower price)", 
                opp.token_pair.symbol1,
                opp.buy_pool.dex
            );
            println!("  3. Swap back to {} on {} (at higher price)", 
                opp.token_pair.symbol0,
                opp.sell_pool.dex
            );
            println!("  4. Repay flash loan + fees");
            println!("  5. Keep profit: ${:.2}", opp.net_profit_usd);
        }
    }
}